### DISCUSSION TITLE
`cloud-init` as backend for instance provisioning

### DISCUSSION THREAD

**Discussion by aavarghese** [skypilot-org/skypilot#3658] [General]
Wondering in the context of similar tools, has skypilot looked into using [clount-init](https://cloud-init.io/) as backend provisioner? Since it is a popular large project used across industries, we're curious about what are some of the contemplated pros/cons for using Skypilot vs cloud-init?

**Comment by Michaelvll** [comment]
Thanks for the question @aavarghese! Based on our best knowledge `cloud-init` is a tool for applying initial boot setup to a newly provisioned VM, instead of provisioning the VM.

We use `cloud-init` to apply the initial setup (e.g., private key setup, etc) on some of the supported clouds, such as AWS and Azure. Please check out the links below:
https://github.com/skypilot-org/skypilot/blob/e6ee397c00e579cf0227c14b6837efa56abf7df3/sky/templates/aws-ray.yml.j2#L86-L115
https://github.com/skypilot-org/skypilot/blob/e6ee397c00e579cf0227c14b6837efa56abf7df3/sky/templates/azure-ray.yml.j2#L72

Do you see any potential improvement for our usage of `cloud-init`?

**Reply by starpit** [reply]
hi @Michaelvll thanks for the help with this!

1. i see a gap in ibm cloud support?
2. cloud-init supports things like `bootcmd` and `runcmd`, which allow for e.g. launching the job as part of cloud-init, rather than requiring any extra layers on top. have you considered leveraging these cloud-init capabilities?

thanks,
nick
